x-pack/metricbeat/module/oracle/tablespace/used_and_free_space.go

```diff
@@ -11,30 +11,24 @@ import (
 )
 
 type usedAndFreeSpace struct {
-	TablespaceName string
-	TotalFreeBytes sql.NullInt64
-	TotalUsedBytes sql.NullInt64
-}
-
-func (d *usedAndFreeSpace) hash() string {
-	return d.TablespaceName
-}
-
-func (d *usedAndFreeSpace) eventKey() string {
-	return d.TablespaceName
+	TablespaceName  string
+	TotalSpaceBytes sql.NullInt64
+	TotalFreeBytes  sql.NullInt64
+	TotalUsedBytes  sql.NullInt64
 }
 
 func (e *tablespaceExtractor) usedAndFreeSpaceData(ctx context.Context) ([]usedAndFreeSpace, error) {
-	rows, err := e.db.QueryContext(ctx, "SELECT b.tablespace_name, tbs_size used, a.free_space free FROM (SELECT tablespace_name, sum(bytes) AS free_space FROM dba_free_space GROUP BY tablespace_name) a, (SELECT tablespace_name, sum(bytes) AS tbs_size FROM dba_data_files GROUP BY tablespace_name) b WHERE a.tablespace_name(+)=b.tablespace_name")
+	rows, err := e.db.QueryContext(ctx, `SELECT b.tablespace_name, (b.tbs_size - a.free_space) used, a.free_space free, (SELECT SUM(bytes) FROM DBA_DATA_FILES) + (SELECT SUM(bytes) FROM DBA_TEMP_FILES) AS total_sum FROM (SELECT tablespace_name, SUM(bytes) AS free_space FROM DBA_FREE_SPACE GROUP BY tablespace_name) a RIGHT JOIN (SELECT tablespace_name, SUM(bytes) AS tbs_size FROM DBA_DATA_FILES GROUP BY tablespace_name) b ON a.tablespace_name = b.tablespace_name`)
```

Comment:
Suggestion

Reasons - Optimisation, Handle situation that `a.free_space` becomes NULL and subrtaction operation fails.


```
SELECT 
    b.tablespace_name, 
    (b.tbs_size - NVL(a.free_space, 0)) AS used, 
    NVL(a.free_space, 0) AS free, 
    (SELECT SUM(bytes) FROM DBA_DATA_FILES) + (SELECT SUM(bytes) FROM DBA_TEMP_FILES) AS total_sum
FROM 
    (SELECT tablespace_name, SUM(bytes) AS free_space FROM DBA_FREE_SPACE GROUP BY tablespace_name) a
RIGHT JOIN 
    (SELECT tablespace_name, SUM(bytes) AS tbs_size FROM DBA_DATA_FILES GROUP BY tablespace_name) b 
ON 
    a.tablespace_name = b.tablespace_name;


```

Reply:
@niraj-elastic  Did you explore DBA_TABLESPACE_USAGE_METRICS? This query can be written in a much simpler way.

See: 

1. https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/DBA_TABLESPACE_USAGE_METRICS.html
2. https://alexandruneda.wordpress.com/oracle/free-space-and-dba_tablespace_usage_metrics/
3. https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31451#issuecomment-2024097323

Reply:
Agree with @agithomas i.e., when doing calculations always make sure that you handle nulls with NVL/ COALESCE

Reply:
Here are a few finding which raise concerns while using DBA_TABLESPACE_USAGE_METRIC  tablespace:
    
- If the query is changed it will be using two new tablespaces DBA_TABLESPACE_USAGE_METRIC and DBA_TABLESPACES whose access needs to be provided which will be huge change for current users of Oracle module.
- This query will be more complicated as DBA_TABLESPACE_USAGE_METRIC will give as results in number of blocks and to get result in to bytes, we will have to derive block size from DBA_TABLESPACES table. Then we will have to multiply number of blocks with block size to get final answer.
- Form Oracle version Version 11.2.0.4 and later  `undo tablespaces` are not [being listed](https://support.oracle.com/knowledge/Oracle%20Database%20Products/2642337_1.html) in DBA_TABLESPACE_USAGE_METRIC which are included in current query.
- For total_space field we are currently summing file sizes of all dba_data_file and file sizes of all dba_temp_files  because of which values of total_space metric will  differ, which is obvious from TABLESPACE_SIZE definition in [DBA_TABLESPACE_USAGE_METRIC ](https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/DBA_TABLESPACE_USAGE_METRICS.html#:~:text=If%20the%20tablespace%20contains%20any%20datafiles%20with%20autoextend%20enabled%2C%20then%20this%20column%20displays%20the%20maximum%20size%20to%20which%20the%20tablespace%20can%20grow%2E)and definition of[ total_bytes](https://github.com/elastic/beats/blob/9c9ae3568309bb4c715fe1c25834eafd06f0f82a/x-pack/metricbeat/module/oracle/tablespace/_meta/docs.asciidoc?plain=1#L25) field. If we choose new query than it will be completely different result for our current oracle module users. 
- As total_space differs so consequently free_space also differs as that can be calculated by  [TABLESPACE_SIZE](https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/DBA_TABLESPACE_USAGE_METRICS.html#:~:text=unexpired%20undo%20segments.-,TABLESPACE_SIZE,-NUMBER) - [.USED_SPACE](https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/DBA_TABLESPACE_USAGE_METRICS.html#:~:text=Tablespace%20name-,USED_SPACE,-NUMBER)

I think using `DBA_TABLESPACE_USAGE_METRIC` won't be good idea since we have active users for this module. What do you think ? @shmsr @agithomas

Reply:
@niraj-elastic Thanks for evaluating DBA_TABLESPACE_USAGE_METRIC; appreciate it. Given the valid concerns let's stick to your original query. Thanks. I'll mark this is as resolved.